compiler/optimizer/abstractinterpreter/InliningProposal.hpp

```diff
@@ -0,0 +1,116 @@
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#ifndef INLINING_PROPOSAL_INCL
+#define INLINING_PROPOSAL_INCL
+
+#include "env/Region.hpp"
+#include "optimizer/abstractinterpreter/IDT.hpp"
+#include "optimizer/abstractinterpreter/IDTNode.hpp"
+#include "infra/BitVector.hpp"
+#include "compile/Compilation.hpp"
+
+namespace TR {
+
+/**
+ *Inlining Proposal records the set of IDTNodes selected to be inlined.
+ */
+class InliningProposal
+   {
+   public:
+
+   InliningProposal(TR::Region& region, TR::IDT *idt);
+   InliningProposal(const TR::InliningProposal&, TR::Region& region);
+
+   void print(TR::Compilation *comp);
+   bool isEmpty();
+
+   uint32_t getCost();
+   double getBenefit();
+
+   /**
+    * @brief add an IDTNode selected to be inlined to the proposal
+    *
+    * @param node the IDTNode
+    */
+   void addNode(TR::IDTNode* node);
+
+   /**
+    * @brief Check if the node is selected to be inlined
+    *
+    * @param node the IDTNode to be checked
+    *
+    * @return true if in proposal. false otherwise.
+    */
+   bool isNodeInProposal(TR::IDTNode* node );
+
+   /**
+    * @brief Union two proposals.
+    *
+    * @param a one proposal
+    * @param b another proposal
+    */
+   void unionInPlace(TR::InliningProposal *a, TR::InliningProposal* b);
+
+   /**
+    * @brief Check if self intersects with another proposal.
+    *
+    * @param other the proposal
+    *
+    * @return true of they intersect. false otherwise.
+    */
+   bool intersects(TR::InliningProposal* other);
+
+   private:
+
+   void computeCostAndBenefit();
+   void ensureBitVectorInitialized();
+
+   TR::Region& _region;
+   TR_BitVector *_nodes;
+   uint32_t _cost;
+   double _benefit;
+   TR::IDT *_idt;
+};
+
+
+class InliningProposalTable
+   {
+   public:
+
+   InliningProposalTable(uint32_t rows, uint32_t cols, TR::Region& region);
+   TR::InliningProposal* get(uint32_t row, uint32_t col);
+   void set(uint32_t row, uint32_t col, TR::InliningProposal* proposal);
+
+   TR::Region& region() { return _region; }
+
+   TR::InliningProposal* getEmptyProposal() { return new (region()) TR::InliningProposal(region(), NULL); }
```

Comment:
Same as https://github.com/eclipse/omr/pull/5509#discussion_r1376440845. I will take a look later to see if this part can be improved.

Reply:
I checked all usage of this `getEmptyProposal()`. It is used by `TR::InliningProposalTable::get()`, and its returned inlining proposal will be used as operand of merge proposal, check if the proposal is empty, check if the given node in the proposal and get the benefit of this proposal. All actions won't make any changes to the empty proposal, so we can reuse only one instance. However, I still prefer ti allocate a new one every time, because we don't know if any future contributor will amend the returned value of `getEmptyProposal()`, and if they amend it, this may cause a wrong inliner packing result.

Reply:
Mutating the result of `TR::InliningProposalTable::get()` would be a bug in general, since it would effectively update the table in-place, and the table entries are supposed to remain unchanged once set. Instead of defensively allocating new empty proposals, could we make sure to catch and prevent such mutations? e.g. maybe proposals could be "frozen" when inserted into the table, and we could fatal assert that they're not frozen when trying to modify them. With something like that, we could more confidently reuse a single empty proposal

Reply:
I have added a private member `_frozen` for `TR::InliningProposal` class, and added some assertions it make sure a `TR::InliningProposal` object won't be mutated when `_frozen` is `true`. The member `_frozen` will also be set as true when the propsal is inserted into a table.

I have also amended  `getEmptyProposal()` to only initialize one static proposal object, and it will also be set as frozen during initialization.

Reply:
Thanks!

Reply:
Thanks!